### DISCUSSION TITLE
Hide empty folder in Customview

### DISCUSSION THREAD

**Discussion by Julia-SOLIGNAC** [pimcore/pimcore#14116] [Q&A / Support]
Hey everyone, I hope you're having a good day.

I want to hide every empty folder in my custoview : 
![image](https://user-images.githubusercontent.com/108328137/213703070-37c34331-aa42-4b0c-a6b8-eb52109697e9.png)
But keep the ones that contains objects of course.

Is there a where clause or something that could help me with that ? 

Thank you :)

**Comment by SamyMP** [comment]
I think you could do that using an event listener. [This one](https://github.com/pimcore/pimcore/blob/48fc7771c98dfbd5e8865aa238065ea868ffb5b3/lib/Event/AdminEvents.php#L176) should do the trick. You would have to check if the dataobjects of type folder have children other than empty folders, and remove them from the list if they don't

**Reply by Julia-SOLIGNAC** [reply]
Thank you I will try this :) But it'll also change the behavior of the classic dataobject tree right ?
And do you know if there is a way to put variables into customviews, like via services.yaml ?

**Reply by SamyMP** [reply]
It will indeed change the behavior, the API won't return empty folders at all, so you might not want to do that if you still want to add objects empty folder via UI, I should have specified that

The other option is via context menu, with a javascript event, by adding an option in the context menu when right-clicking with `pimcore.events.prepareObjectTreeContextMenu`.

 ```js
 document.addEventListener(pimcore.events.prepareObjectTreeContextMenu, (e) => {
    const { detail } = e;
    if (detail.object.data.type === "folder" && detail.object.childNodes.length === 0) {
        detail.menu.add(new Ext.menu.Item({
            text: "Hide folder",
            iconCls: "pimcore_icon_hide",
            handler: function () {
                detail.object.set('visible', false);
                // You could also code a logic to recursively hide every empty folders, 
                // not just the one you clicked
            }
        }));

        detail.menu.add(new Ext.menu.Item({
            text: "Show folder",
            iconCls: "pimcore_icon_seemode",
            handler: function () {
                detail.object.set('visible', true);
                // Same as above
            }
        }));
    }
})
 ```

I don't know if there is a way to pass variables to customviews, sorry

**Reply by Julia-SOLIGNAC** [reply]
Indeed, the first solution don't fit my need. But I will see if the second one can help me. Thank you very much for your assistance 👍
